d4n topic branch

pritha-srivastava opened the pull request:
So far, this PR has achieved a working local read path. To test it, bring up vstart as follows (assuming that ssd cache backend is being used as the cache driver)
MON=1 OSD=1 RGW=1 MGR=0 MDS=0 ../src/vstart.sh -n -d -o rgw_enable_ops_log=true -o rgw_d4n_l1_datacache_persistent_path=`<path to datacache>` -o rgw_d4n_l1_datacache_size=5368709120 -o rgw_filter=d4n
1. Create an object of size 30 MB.
2. Create a bucket - s3cmd mb s3://my-bucket
3. Upload the object to the bucket - s3cmd put ./30M s3://my-bucket
4. Get the object from the bucket - s3cmd get s3://my-bucket/30M /dev/shm/30M-dat --force (Check the logs to see that the object has been fetched from the storage backend)
5. Check datacache directory - must have data cached
6. Remove file downloaded in 4. to ensure that the file downloaded in 7. is from the cache - rm -rf /dev/shm/30M-dat
7. Get the object from the bucket - s3cmd get s3://my-bucket/1M /dev/shm/30M-dat --force (Check the logs to see that the object has now been fetched from the cache backend)
8. Use md5 to ensure that the contents of the file downloaded from the cache is correct - md5sum ./30M /dev/shm/30M-dat

TODO:
1. free space calculation of redis backed cache backend.
2. Add orchestration for flushing the redis server before any workflow runs. **Note:** Needed for vstart but we may not want to flush the redis server while running code in production.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+        block.size = bl.length();
+        block.blockId = oid;
+        block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port)); //check if this is the correct way to get the local addr
+        block.size = source->get_obj_size();
```
Redeclaration of `block.size`

pritha-srivastava replied:
Yes - block.size must *NOT* be source->get_obj_size();, please remove this in your PR.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+      ofs += bl_len;
+      block.blockId = oid;
+      block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port)); //check if this is the correct way to get the local addr
+      block.size = source->get_obj_size();
```
here block.size must be bl.length()

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+      bl.splice(0, len_to_copy, &bl_copy);
+      bl_rem.claim_append(bl_copy);
+      block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port));
+      block.size = source->get_obj_size();
```
@samarahu : block.size must be bl_rem.length(), please change it.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+      block.cacheObj.objName = source->get_key().get_oid();
+      ret = filter->get_cache_driver()->put_async(save_dpp, oid, bl, bl.length(), source->get_attrs());
+      if (ret == 0) {
+        block.size = bl.length();
```
@samarahu : actually the size was set here correctly, the previous statement needs to be just removed.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,64 +396,506 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+
+        ret = filter->get_cache_driver()->put_async(save_dpp, oid, bl_rem, bl_rem.length(), source->get_attrs());
+        if (ret == 0) {
+          block.size = bl_rem.length();
```
@samarahu : same as above, the size was set here correctly, the previous statement needs to be removed

pritha-srivastava replied:
Anyway, I will update this PR and rebase with latest main.

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+  ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << "oid: " << oid << " ofs: " << ofs << " end: " << end << dendl;
+
+  this->client_cb = cb;
+  this->cb->set_client_cb(cb, dpp); // what's this for? -Sam
```
What is the difference between `D4NFilterReadOp::client_cb` and `D4NFilterGetCB::cb`?

pritha-srivastava replied:
There are two callbacks here:
1. One the iterate method receives from the upper layer to send back data read.
2. The second callback belongs to the Filterdriver.
Both are explained below:
`cb` is the callback that we receive from the upper layer. You can see https://github.com/ceph/ceph/blob/main/src/rgw/rgw_op.cc#L2205 and the callback is being passed to the `iterate` method here: https://github.com/ceph/ceph/blob/main/src/rgw/rgw_op.cc#L2416. `cb` is being saved in `this->client_cb` in `D4NFilterObject::D4NFilterReadOp`. The callback method of the filter driver is `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data`, which is stored in `this->cb`. `this->cb` is passed to the backend store, so that when data is sent back from the store, the filter driver callback  intercepts it and writes it to the cache (as you can see in `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data`). Now the same data needs to be sent back to the upper layer/client also, so the upper layer callback that we saved in `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB` with the help of `this->cb->set_client_cb(cb, dpp);` is invoked in `D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data` in `auto r = client_cb->handle_data(bl, bl_ofs, bl_len);` Now `cb` saved in `D4NFilterObject::D4NFilterReadOp` is invoked, when data is present in the cache and is read from it - this data is returned back to the upper layer/client in `D4NFilterObject::D4NFilterReadOp::flush`, in this line : `int r = client_cb->handle_data(bl, 0, bl.length());`

samarahu replied:
I understand now; thank you for the explanation!

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+    ofs = adjusted_start_ofs; // redundant? -Sam
+  }
+
+  this->cb->set_ofs(ofs);
```
If the condition on line 638 is met, then `this->cb->set_ofs(ofs)` is called twice. Is this intended?

pritha-srivastava replied:
Well, this piece of code can definitely be reorganised, will do so, thanks

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+    const std::lock_guard l(d4n_get_data_lock);
+    rgw::d4n::CacheBlock block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    block.hostsList.push_back(blockDir->get_addr().host + ":" + std::to_string(blockDir->get_addr().port));
```
Should the entire object also be stored in the local cache or only the block?

samarahu replied:
Additionally, the entire block must be filled out correctly by adding accurate values for the creation time, version, and dirty bit.

pritha-srivastava replied:
We are storing blocks in case of a read cache. When we have the remote caches plugged, all the blocks will be distributed to different caches available. Currently we are caching all blocks belonging to the same object in the same cache. If we read the whole object from the backend, we will break it up into blocks and store them in the cache. If we read partial data from the backend (in case of ranged requests), we will save that in the cache

pritha-srivastava replied:
> Additionally, the entire block must be filled out correctly by adding accurate values for the creation time, version, and dirty bit. I will work on this but the version is still not set up

Correct, I have added whatever I could. You are free to fill the other variables - I can see that you have already done that in your PR. I am not sure about the `dirty bit` though, since that applies to the `write-back` cache, which will have data to be written to the backend store.

samarahu replied:
I understand that the blocks that are chunked will be stored in the cache, but I am wondering if I should also update the cache object's host list to store the local cache address. I assume I shouldn't do this because `put_async` is only being called on the block and not the entire object. The object will be handled separately elsewhere, correct?

samarahu replied:
In regards to the `dirty` bit, I'm also not certain how that will be used but since the write and read caches are being unified, I will keep it for the write cache to use. I can also ask how this should be handled in the read workflow, if at all. Do we plan to have the `version` feature in this PR? If so, I can ask Casey more on implementation details.

samarahu replied:
Update: the `dirty` bit will not be used in the read workflow, so it will simply have a default value of `false` when being set here.

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+      }
+      if (filter->get_cache_driver()->put_async(dpp, oid, bl, bl.length(), source->get_attrs()) == 0) {
+        filter->get_policy_driver()->cachePolicy->update(dpp, oid, ofs, bl.length(), filter->get_cache_driver());
+        /* Store block in directory */
```
If the block exists, shouldn't we be updating it? For example, we should append the local cache address to its host list.

pritha-srivastava replied:
Yes, if the block exists in the directory, we will update the hosts list with the local host (again this applies when remote caches are available), you have done that in your PR and thats fine.

pritha-srivastava replied:
Additionally, we shouldn't be storing any attrs with blocks here. The idea is to fetch the attrs from the backend store for the object, instead of storing all the attrs with each block. We don't cache the head object (which contains the attrs). This needs to be fixed.

samarahu replied:
Additionally, if the block exists, should the rest of the block's attributes be updated in addition to the host list? For example, the block ID and length.

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+      block.blockId = ofs;
+      block.size = bl.length();
+      uint64_t freeSpace = filter->get_cache_driver()->get_free_space(dpp);
+      while(freeSpace < block.size) {
```
Are we only checking for the block size and not adding the attribute size here because the attributes are being stored for the entire object rather than just the block?

pritha-srivastava replied:
Correct. But are you concerned about the `localWeight` attribute that gets added as part of the `update` call?

pritha-srivastava replied:
But we had decided to move this check to `eviction` call correct?

samarahu replied:
The available cache space is calculated in the `eviction` call and yes, the `localWeight` attribute will also take up a slight amount of space that must be considered during eviction. This should account for both the attribute name and value size, correct? With the attribute name including the `user.rgw.` prefix?

samarahu replied:
Update: Due to a recent discussion on this subject, we will keep the block size as the size to free up in the `eviction` call.

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -352,60 +404,501 @@ int D4NFilterObject::D4NFilterReadOp::prepare(optional_yield y, const DoutPrefix
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << 
+    " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+
+    if (source->driver->get_policy_driver()->cachePolicy->exist_key(oid_in_cache)) { 
```
When there is support for communicating with remote caches, we must add a check here for that and I will implement the final piece of the LFUDA algorithm.

samarahu commented on src/rgw/driver/d4n/d4n_directory.h:
```diff
@@ -1,53 +1,101 @@
+
+    void init(CephContext* _cct) {
+      cct = _cct;
+      addr.host = cct->_conf->rgw_d4n_host;
```
Will the object directory always have the same address as the block directory or should there be a config option for each directory type?

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -0,0 +1,421 @@
+    return nullptr;
+
+  /* Get victim cache block */
+  std::string key = entries_heap.top()->key;
```
@samarahu : I am not sure whether we resolved my comment about two entries/ blocks having the same `weight`?

samarahu replied:
Yes; Orran said that, if this happens, the block is just randomly picked. There is no need for LRU here

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -0,0 +1,430 @@
+      return -1;
+    }
+
+    std::string key = victim->cacheObj.bucketName + "_" + victim->cacheObj.objName + "_" + std::to_string(victim->blockID) + "_" + std::to_string(victim->size);
```
@samarahu : I noticed that the `key` here doesn't match with the way `key` that is passed in the `update` method, so this will always result in the key not being found. P.S: The key names now include `version` also.

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -0,0 +1,430 @@
+  int localWeight = age;
+  auto entry = find_entry(key);
+  if (entry != nullptr) { 
+    entry->localWeight += age;
```
@samarahu : shouldn't localWeight increase by 1 on every access, before adding cache age to it? else localWeight of all blocks will always remain 0, the initial value of cache age is 0, and while eviction, the new cache age is based on the local weight of the evicted block and the cache age, both of which are 0, so the new cache age will also be zero.

pritha-srivastava replied:
@samarahu : did you address the above comment

samarahu replied:
I updated this so that `entry->localWeight` is not altered before the `erase` call. Instead, a copy is stored in the `localWeight` variable declared on line 331 and has the `age` added to it. It is later used to set the updated `localWeight` for the new entry.

pritha-srivastava commented:
@cbodley @adamemerson : can you please look at the merge conflicts in the neorados part in this PR, and tell me which changes should go in?

cbodley commented:
> @cbodley @adamemerson : can you please look at the merge conflicts in the neorados part in this PR, and tell me which changes should go in?

yeah there's a ton of stuff from one of Adam's old branches that we don't need here. the only parts i *know* we need are access to the global io_context from commits https://github.com/ceph/ceph/pull/52444/commits/50448ce939ac04cfc52552e42a67311c3d880ba2 https://github.com/ceph/ceph/pull/52444/commits/8f54f5b148fad154eb1ea73cad4e0234139dbccd https://github.com/ceph/ceph/pull/52444/commits/349deedd601be32623da55fb9e2461253dc0fe93

i also see a merge commit https://github.com/ceph/ceph/pull/52444/commits/6a60dbc308b37f7f4aacc0b6f84d9518af9db017 in this branch, which we'll need to remove before merging

pritha-srivastava commented:
> > @cbodley @adamemerson : can you please look at the merge conflicts in the neorados part in this PR, and tell me which changes should go in?
> 
> yeah there's a ton of stuff from one of Adam's old branches that we don't need here. the only parts i _know_ we need are access to the global io_context from commits [50448ce](https://github.com/ceph/ceph/commit/50448ce939ac04cfc52552e42a67311c3d880ba2) [8f54f5b](https://github.com/ceph/ceph/commit/8f54f5b148fad154eb1ea73cad4e0234139dbccd) [349deed](https://github.com/ceph/ceph/commit/349deedd601be32623da55fb9e2461253dc0fe93)
> 
> i also see a merge commit [6a60dbc](https://github.com/ceph/ceph/commit/6a60dbc308b37f7f4aacc0b6f84d9518af9db017) in this branch, which we'll need to remove before merging

@cbodley : should I remove all other commits related to neorados and have only 50448ce, 8f54f5b, and 349deed.

samarahu commented on src/rgw/driver/d4n/d4n_policy.h:
```diff
@@ -0,0 +1,169 @@
+    virtual int eviction(const DoutPrefixProvider* dpp, uint64_t size, optional_yield y) override;
+    virtual void update(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, std::string version, optional_yield y) override;
+    virtual bool erase(const DoutPrefixProvider* dpp, const std::string& key, optional_yield y) override;
+    virtual void shutdown() override;
```
Make specific to LFUDA

pritha-srivastava commented on src/rgw/rgw_redis_driver.h:
```diff
@@ -0,0 +1,86 @@
+  protected:
+    std::shared_ptr<connection> conn;
+
+    static std::unordered_map<std::string, Partition> partitions;
+    Partition partition_info;
+    uint64_t free_space;
+    uint64_t outstanding_write_size;
+
+    int add_partition_info(Partition& info);
+    int remove_partition_info(Partition& info);
```
We will remove this, as discussed in the meeting

pritha-srivastava commented on src/rgw/rgw_ssd_driver.cc:
```diff
@@ -0,0 +1,566 @@
+    if (attr_size < 0) {
+        auto ret = errno;
+        ldpp_dout(dpp, 0) << "ERROR: could not get attribute " << attr_name << ": " << cpp_strerror(ret) << dendl;
+        return "";
```
this case to handle errors, the condition when `attr_size == 0` is handled below, but yes, I will change the return value to indicate success or error.

pritha-srivastava commented:
I have addressed some review comments here: https://github.com/pritha-srivastava/ceph/tree/wip-rgw-d4n-review-comments, will cherry-pick them once we have a resolution for the rebase issues.

pritha-srivastava commented:
@cbodley : The test plan for the read workflow is to write a workunit similar to that for d3n: https://github.com/ceph/ceph/blob/main/qa/workunits/rgw/test_rgw_datacache.py. The test would create a file of a particular size and then upload it to a bucket. `get` would be invoked to get the file. A second `get` would be invoked to get the file from the cache. There would be checks to see whether the cache directory has been populated with files. I can see that a workunit already exists for d4n, we could probably add to that. (and override the config variables in the correct yaml files). Is this acceptable?

pritha-srivastava commented:
@cbodley - This branch is rebased, cleaned up and with the review comments incorporated starting from commit b7e4bd6870b92665e6abc57f4a6e86a88f1652cc

pritha-srivastava commented:
@samarahu : I was able to successfully push this branch to shaman to build - should show up as `wip-rgw-d4n`

pritha-srivastava commented:
> @samarahu : I was able to successfully push this branch to shaman to build - should show up as `wip-rgw-d4n`

Ok, I see the same failure related to credentials - `Failed Publishing GitHub checks: java.io.IOException: Unsupported credential type: com.cloudbees.jenkins.plugins.sshcredentials.impl.BasicSSHUserPrivateKey
Finished: FAILURE`

samarahu commented on src/rgw/rgw_redis_driver.h:
```diff
@@ -0,0 +1,86 @@
+#include "rgw_common.h"
+#include "rgw_cache_driver.h"
+
+#define dout_subsys ceph_subsys_rgw
+#define dout_context g_ceph_context
```
I will move them to the .cc file

pritha-srivastava commented:
@samarahu : I built this branch today and see this failure:
```Running submodule update ...
+ git submodule update --init
fatal: No url found for submodule path 'build/boost_redis' in .gitmodules
```
followed by the credentials error in the end:
```{}[GitHub Checks] Failed Publishing GitHub checks: java.io.IOException: Unsupported credential type: com.cloudbees.jenkins.plugins.sshcredentials.impl.BasicSSHUserPrivateKey
Finished: FAILURE
```

Has any commit related to build/redis removal been missed - I have used Adam's fixed branch for reference - https://github.com/ceph/ceph/pull/54976, please check and let me know.

samarahu commented:
> @samarahu : I built this branch today and see this failure:
> 
> ```
> Running submodule update ...
> + git submodule update --init
> fatal: No url found for submodule path 'build/boost_redis' in .gitmodules
> ```
> 
> followed by the credentials error in the end:
> 
> ```
> {}[GitHub Checks] Failed Publishing GitHub checks: java.io.IOException: Unsupported credential type: com.cloudbees.jenkins.plugins.sshcredentials.impl.BasicSSHUserPrivateKey
> Finished: FAILURE
> ```
> 
> The build log is here: https://jenkins.ceph.com/job/ceph-dev-new-setup/76626//consoleFull Has any commit related to build/redis removal been missed - I have used Adam's fixed branch for reference - #54976, please check and let me know. Check [dac8575](https://github.com/ceph/ceph/commit/dac85757e69d43b0f4fc2da6c3ffd89d9cab9a02) on this branch
> 
> I have also scheduled a build for a completely new branch wip-prsrivas-rgw-d4n off this branch in shaman

@pritha-srivastava I ran into the same error locally but I resolved it by running `git rm --cached ../../build/boost_redis/` in the `src` directory. I have looked over the commit you've linked and ensured that nothing from my end is missing.

pritha-srivastava commented:
> > @samarahu : I built this branch today and see this failure:
> > ```
> > Running submodule update ...
> > + git submodule update --init
> > fatal: No url found for submodule path 'build/boost_redis' in .gitmodules
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > followed by the credentials error in the end:
> > ```
> > {}[GitHub Checks] Failed Publishing GitHub checks: java.io.IOException: Unsupported credential type: com.cloudbees.jenkins.plugins.sshcredentials.impl.BasicSSHUserPrivateKey
> > Finished: FAILURE
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The build log is here: https://jenkins.ceph.com/job/ceph-dev-new-setup/76626//consoleFull Has any commit related to build/redis removal been missed - I have used Adam's fixed branch for reference - #54976, please check and let me know. Check [dac8575](https://github.com/ceph/ceph/commit/dac85757e69d43b0f4fc2da6c3ffd89d9cab9a02) on this branch
> > I have also scheduled a build for a completely new branch wip-prsrivas-rgw-d4n off this branch in shaman
> 
> @pritha-srivastava I ran into the same error locally but I resolved it by running `git rm --cached ../../build/boost_redis/` in the `src` directory. I have looked over the commit you've linked and ensured that nothing from my end is missing.

@samarahu : I noticed that even successful builds in shaman have the error:
```
[GitHub Checks] Failed Publishing GitHub checks: java.io.IOException: Unsupported credential type: com.cloudbees.jenkins.plugins.sshcredentials.impl.BasicSSHUserPrivateKey
```
So I believe that is not the real problem and if there is a build failure we should check the log file to figure out the actual cause of the failure.

pritha-srivastava commented:
@samarahu : I have been debugging the `get` call hanging for larger objects - 30M, and although I observe the hang in different places - this is one place where I saw the code hanging:
```
(gdb) bt
#0  futex_wait (private=0, expected=2, futex_word=0x5604dc873028) at ../sysdeps/nptl/futex-internal.h:146
#1  __GI___lll_lock_wait (futex=futex@entry=0x5604dc873028, private=0) at lowlevellock.c:49
#2  0x00007f1db21cee41 in lll_mutex_lock_optimized (mutex=0x5604dc873028) at pthread_mutex_lock.c:48
#3  ___pthread_mutex_lock (mutex=0x5604dc873028) at pthread_mutex_lock.c:93
#4  0x00005604d4e7da37 in __gthread_mutex_lock (__mutex=0x5604dc873028) at /usr/include/c++/13/x86_64-redhat-linux/bits/gthr-default.h:749
#5  std::mutex::lock (this=this@entry=0x5604dc873028) at /usr/include/c++/13/bits/std_mutex.h:113
#6  0x00005604d5580b8b in std::lock_guard<std::mutex>::lock_guard (__m=..., this=0x7f1c7e346068) at /usr/include/c++/13/bits/std_mutex.h:249
#7  rgw::d4n::LFUDAPolicy::exist_key (this=this@entry=0x5604dc872fc0, key="my-bucket_d6c6583d-7791-4383-b7ac-6036c45be58a.4153.12251551462825885513_30M_0_4194304") at /home/prsrivas/ceph/src/rgw/driver/d4n/d4n_policy.cc:203
#8  0x00005604d526729b in rgw::sal::D4NFilterObject::D4NFilterReadOp::iterate (this=0x5604df1b8320, dpp=<optimized out>, ofs=<optimized out>, end=<optimized out>, cb=<optimized out>, y=...)
    at /home/prsrivas/ceph/src/rgw/driver/d4n/rgw_sal_d4n.cc:560
#9  0x00005604d50c28b4 in RGWGetObj::execute (this=0x5604e0ccb340, y=...) at /home/prsrivas/ceph/src/rgw/rgw_op.cc:2457
```

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -0,0 +1,445 @@
+    return -EINVAL;
+  }
+
+  if (!std::get<0>(resp).value()) {
```
@samarahu : As part of debugging the hanging `get` call, I discovered that the value of `std::get<0>(resp).value()` is always 1. I believe, the initial value should be 0, and this method will be invoked as part of `update` call, and then the initial local_weight_sum will be set using `sum` calculated below. But that doesn't happen and EINVAL is returned from the next try catch loop, and from this point the local weight sum calculation goes wrong.
Please correct me if I am wrong.

samarahu replied:
I've confirmed that the `set_local_weight_sum` call succeeds even when there is hanging behavior, so there is no EINVAL return here and the method completes successfully, yet the op still hangs. I don't believe this is the source of the bug

pritha-srivastava replied:
I am not saying this is the source of the bug, I came across this while debugging the issue. I can share logs with you to take a look at.

pritha-srivastava replied:
Note: Run `redis-cli FLUSHALL` to flush redis server of all existing values - this solves the issue.

pritha-srivastava replied:
Also as discussed - please modify the argument type of `weight` appropriately in `set_local_weight_sum`

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
 
-  return next->delete_obj(dpp, y);
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << "prefix: " << prefix << dendl;
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << "oid: " << source->get_key().get_oid() << " ofs: " << ofs << " end: " << end << dendl;
```
Are these meant to log "D4NFilterObject" rather than "D3NFilterObject"?

pritha-srivastava commented:
@samarahu @cbodley @mattbenjamin @adamemerson : I have been debugging the `get` call hanging issue a bit more - it hangs even after  applying Samarah's change of establishing a single connection across all d4n classes. I enabled verbose logging for redis server and ran the redis-cli to `monitor` `SET`/`GET` commands. Again as before I consistently see that the command hangs due to `Operation Cancelled` error or -125 error. The redis monitor logs and server logs show that once this error is returned, the client initiates a connection with the server on a different port. All this while I thought that the server closed the connection, but the server logs indicate otherwise - the client is resetting the connection. I am attaching snippets of logs below:
1. rgw log file:
```2024-01-12T13:59:03.897+0530 7f0c024916c0 10 req 9373245644651378461 0.012000063s s3:get_obj LFUDAPolicy::update(): Before get_local_weight_sum().
**2024-01-12T13:59:38.392+0530 7f0bf24716c0 10 req 9373245644651378461 34.507183075s s3:get_obj LFUDAPolicy::update(): localWeights is: .-125**
2024-01-12T14:00:38.393+0530 7f0bec4656c0 10 req 9373245644651378461 94.507499695s s3:get_obj LFUDAPolicy::get_local_weight_sum(): exists: 1
```

2. redis monitor logs, please note the port number changing from 51818 to 44662 and re-establishment of connection , also I had changed the key here just for debugging purpose to lfudaLocaWeights
```1705048143.898164 [0 127.0.0.1:51818] "HEXISTS" "lfudaLocaWeights" "localWeights"
1705048148.397717 [0 127.0.0.1:51818] "PING" "Boost.Redis"
1705048238.394638 [0 127.0.0.1:44662] "HELLO" "3" "SETNAME" "D4N.Filter"
```

3. redis server logs: the connection is re-initiated after 1 minute due to a setting in config.hpp of boost redis, which I had set to 1 minute (reconnect_wait_interval)
```49789:M 12 Jan 2024 13:57:08.392 - Accepted 127.0.0.1:51818
49789:M 12 Jan 2024 13:59:38.393 - Reading from client: Connection reset by peer
49789:M 12 Jan 2024 14:00:38.394 - Accepted 127.0.0.1:44662
```

Does anyone know how to enable boost.redis logging, there is a config option `std::string log_prefix = "(Boost.Redis) ";
` but I couldn't find a way to enable these logs - they could help see what's happening at the client side. Also as a workaround, if we reduce the reconnect_wait_interval, and wait to reconnect when -125 error is returned, the read workflow eventually completes after reconnect_wait_interval. But I believe it is good to know the cause of these -125 errors.

P.S: I will be writing some tests to continue debugging this behaviour.

samarahu commented:
@pritha-srivastava You can change the logging level to `debug` with the following code:
```
using boost::redis::logger;
conn->async_run(cfg, logger::level::debug, net::consign(net::detached, conn));
```

pritha-srivastava commented:
I got the boost.redis logging working finally - And I could see this in the log file for `Operation Cancelled` error:
```
(Boost.Redis) Connection lost: Operation canceled
(Boost.Redis) Resolve results: 127.0.0.1:6379
(Boost.Redis) Connected to endpoint: 127.0.0.1:6379
```

Looking at boost redis code, I could also see this explanation for `async_run`:
```
 4. Start a health-check operation where ping commands are sent
    *     at intervals specified in
    *     `boost::redis::config::health_check_interval`.  The message passed to
    *     `PING` will be `boost::redis::config::health_check_id`.  Passing a
    *     timeout with value zero will disable health-checks.  If the Redis
    *     server does not respond to a health-check within two times the value
    *     specified here, it will be considered unresponsive and the connection
    *     will be closed and a new connection will be stablished.
 ```
Which leads me to think that the connection would have been closed because of the health check not being responded to in the expected time by the server; the client does eventually re-connect and continue like in the code below:
```
conn_->impl_.async_run(conn_->cfg_, logger_, std::move(self));
         conn_->cancel(operation::receive);
         logger_.on_connection_lost(ec);
         if (!conn_->will_reconnect() || is_cancelled(self)) {
            conn_->cancel(operation::reconnection);
            self.complete(!!ec ? ec : asio::error::operation_aborted);
            return;
         }

         conn_->timer_.expires_after(conn_->cfg_.reconnect_wait_interval);
         BOOST_ASIO_CORO_YIELD
         conn_->timer_.async_wait(std::move(self));
         BOOST_REDIS_CHECK_OP0(;)
         if (!conn_->will_reconnect()) {
            self.complete(asio::error::operation_aborted);
            return;
         }
         conn_->reset_stream();
```
So it could be that the `Operation Cancelled` error can be ignored and handled by waiting for the client to re-connect. There is another hang which still needs to be debugged - will add logs in boost redis code and continue.

cbodley commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+    ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << 
+    " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
+
+    if (source->driver->get_policy_driver()->get_cache_policy()->exist_key(oid_in_cache) > 0) { 
+      // Read From Cache
+      auto completed = source->driver->get_cache_driver()->get_async(dpp, y, aio.get(), oid_in_cache, read_ofs, len_to_read, cost, id); 
+
+      source->driver->get_policy_driver()->get_cache_policy()->update(dpp, oid_in_cache, adjusted_start_ofs, part_len, version, y);
```
i'm confused by this pattern, given that `get_async()` is async. shouldn't we wait for this block to complete before calling `update()` on it? the `completed` list here will only contain other blocks that previously completed - you should probably loop over those to make their corresponding `update()` calls

pritha-srivastava replied:
@cbodley : This is on my TODO list - I have been thinking of shifting this to `flush` after the data is sent via `handle_data`. I need to figure out how to have the correct offset, length and version of the block accessible in `flush`. OR maybe simply do it after `flush`.

cbodley replied:
`flush()` seems like a good place, yeah

> I need to figure out how to have the correct offset, length and version of the block

`completed` is a list of [rgw::AioResult](https://github.com/ceph/ceph/blob/5ea5521/src/rgw/rgw_aio.h#L34-L38)s. we put the offset in `AioResult::id` so we can sort the results. you can get the length from `AioResult::data.length()`. for version, i think that should be the same for all of the reads?

pritha-srivastava replied:
Yes version shouldn't be a problem, but for ranged requests - the offset and length of the block cached might be different from the offset and length of data returned to the end user

cbodley commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
 
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d4n_get_data_lock);
```
this lock has a huge scope; avoid holding mutexes over i/o like get_cache_driver()->put_async() and get_cache_policy()->update()

pritha-srivastava replied:
Yes, checking all the locks is also on my TODO list

cbodley replied:
within a single `D4NFilterReadOp`, i don't think you need that mutex at all

the AioThrottle from `aio = rgw::make_throttle()` should handle any thread-safety issues for you (where io completions happen in a different background thread), as long as you're only accessing `aio` in `D4NFilterReadOp::iterate()`

pritha-srivastava replied:
the lock here is for writing to the cache backend - wouldn't that be needed if we have two parallel `get` calls, and both of them read from the backend store and then try to write to the backend? Can such a scenario not arise?

cbodley commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+        block.version = version;
+        auto ret = filter->get_policy_driver()->get_cache_policy()->eviction(dpp, block.size, *y);
+        if (ret == 0) {
+          //Should we replace each put_async with put, to ensure data is actually written to the cache before updating the data structures and before the lock is released?
+          ret = filter->get_cache_driver()->put_async(dpp, oid, bl, bl.length(), attrs);
+          if (ret == 0) {
+            filter->get_policy_driver()->get_cache_policy()->update(dpp, oid, ofs, bl.length(), version, *y);
```
yeah, there should probably be a version of put() that takes optional_yield so we can wait for completion here

pritha-srivastava replied:
@cbodley : Can't we just use the blocking  `put` here, the one which is non-asynchronous. It does take in `yield` but doesn't do anything with it.

cbodley replied:
sure. i see that `RedisDriver::put()` handles yield correctly. SSDDriver should eventually support it too, using logic similar to the CompletionToken-templated overload of `SSDDriver::get_async()`

pritha-srivastava commented:
jenkins retest this please

pritha-srivastava commented:
jenkins test api

samarahu commented:
Green run of the D4N suite against this branch: https://pulpito.ceph.com/suriarte-2024-02-26_19:17:31-rgw:d4n-wip-rgw-d4n-distro-default-smithi/

pritha-srivastava commented:
@cbodley , @mattbenjamin , @dang , @alimaredia : Request all of you to leave additional comments if there are any. Also Samarah has completed writing teuthology tests for large and small objects with d4n filter driver enabled and has been able to get a green teuthology run. Is there anything else that we need to address to get this PR merged?

Samarah has been working on some parts of the Redis driver, but i believe that shouldn't block this PR from getting merged as the local read workflow currently uses the ssd cache backend.

cc @samarahu

pritha-srivastava commented:
We have an almost clean run here: https://pulpito.ceph.com/suriarte-2024-03-04_17:13:34-rgw-wip-rgw-d4n-distro-default-smithi/

The multisite failures are there on main as well, I have checked each test case failing.
6 jobs are dead, and I have scheduled another run here: https://pulpito.ceph.com/prsrivas-2024-03-05_05:47:02-rgw-wip-rgw-d4n-distro-default-smithi/

pritha-srivastava commented:
The latest run is here: https://pulpito.ceph.com/prsrivas-2024-03-06_05:23:10-rgw-wip-rgw-d4n-distro-default-smithi/, There are two multisite failures and a third failure - `sudo fuser -v /var/lib/dpkg/lock-frontend` which looks like an installation issue in one of the rgw upgrade tests.

pritha-srivastava commented:
jenkins test make check

pritha-srivastava commented:
@cbodley @mattbenjamin @dang : The latest run is here with only multisite failures: https://pulpito.ceph.com/suriarte-2024-03-06_18:36:43-rgw-wip-rgw-d4n-distro-default-smithi/

also make check is failing because of a failure in run-tox-qa and ceph API tests show a failure in dashboard.

mattbenjamin commented:
I hope @pritha-srivastava 's reported results mean we can now merge this?

pritha-srivastava commented:
jenkins test make check

pritha-srivastava commented:
jenkins test make check arm64

pritha-srivastava commented:
jenkins test api

pritha-srivastava commented:
jenkins test make check

pritha-srivastava commented:
@cbodley , @mattbenjamin : even after rebasing this branch onto main, I see a dashboard related failure in ceph API tests (this was passing before the rebase) and run-tox-qa test failing in make check. The rebase has fixed make check(arm64). Any idea how to get past this? I have posted about this in ceph slack.

nizamial09 commented:
This is what i see in the logs of [make check](
https://jenkins.ceph.com/job/ceph-pull-requests/131134/console)
```
flake8: commands[0] /home/jenkins-build/build/workspace/ceph-pull-requests/qa> flake8 --select=F,E9 --exclude=venv,.tox
./tasks/d4ntests.py:5:1: F401 'teuthology.orchestra.run' imported but unused
./tasks/redis.py:5:1: F401 'teuthology.orchestra.run' imported but unused
flake8: exit 1 (3.15 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/qa> flake8 --select=F,E9 --exclude=venv,.tox pid=303091
flake8: FAIL ✖ in 14.15 seconds
```
And I believe that might be related to your changes since you are modifying these files.

API test failures are known and i havent been able to find the real cause of it since it doesnt appear everytime consistently. But I havent debugged it deeper to say more.

pritha-srivastava commented:
jenkins test api

pritha-srivastava commented:
jenkins test windows

pritha-srivastava commented:
jenkins test make check arm64

samarahu commented:
jenkins test api

samarahu commented:
RGW Suite test results: https://pulpito.ceph.com/prsrivas-2024-03-15_10:45:14-rgw-wip-rgw-d4n-distro-default-smithi/
The 3 failures are known and expected.

pritha-srivastava commented:
@cbodley @mattbenjamin : we have a green run in teuthology (with known failures), and the ceph api tests and ceph windows tests failures do not look related to this PR, the error in ceph API test is `‘/home/jenkins-build/build/workspace/ceph-api/build/out’: No such file or directory` and in ceph windows test is:
`+ [[ ! -z /home/jenkins-build/build/workspace/ceph-windows-pull-requests/libvirt/known_hosts ]]
+ SSH_OPTS=' -o UserKnownHostsFile=/home/jenkins-build/build/workspace/ceph-windows-pull-requests/libvirt/known_hosts'
+ timeout 30m ssh -i **** -o UserKnownHostsFile=/home/jenkins-build/build/workspace/ceph-windows-pull-requests/libvirt/known_hosts ubuntu@192.168.122.16 ./ceph-vstart.sh
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
Error initializing cluster client: ObjectNotFound('RADOS object not found (error calling conf_read_file)')
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH **
`

pritha-srivastava commented:
jenkins test api

pritha-srivastava commented:
jenkins test windows

samarahu commented:
jenkins test windows

samarahu commented:
jenkins test api

samarahu commented:
@cbodley @mattbenjamin This PR passes all checks and has been tested on Teuthology. Are there any remaining tasks to complete before merging?

cbodley commented:
a different version of Adam's commit [rgw: Thread site config into RADOS driver](https://github.com/ceph/ceph/pull/52444/commits/e81fdd31bc3c87b3de5ba910338e86320861684b) already merged to main, please remove that one

alimaredia commented:
The two follow up threads of work I think should happen are:
1. Docs on how D4N works and more detailed descriptions of the D4N related config variables
2. Entertain the possibility of compiling without D4N

pritha-srivastava commented:
> a different version of Adam's commit [rgw: Thread site config into RADOS driver](https://github.com/ceph/ceph/pull/52444/commits/e81fdd31bc3c87b3de5ba910338e86320861684b) already merged to main, please remove that one

@cbodley , @mattbenjamin , @alimaredia , @dang : rebasing didn't remove Adam's commit, I didn't get a merge conflict either - I have pushed the locally rebased branch here: https://github.com/ceph/ceph/compare/main...pritha-srivastava:ceph:wip-rgw-d4n-test. Should I just remove that commit from the branch?

alimaredia commented:
jenkins test make check arm64

pritha-srivastava commented:
jenkins test make check arm64

pritha-srivastava commented:
@cbodley @alimaredia @mattbenjamin @dang : make check arm64 fails repeatedly with the same errors - `Unable to locate package ceph-libboost-atomic1.82-dev` and a failure in `unittest_mempool`. I have seen these errors in other branches of mine as well. I have asked about this in the sepia channel, but does anyone have any suggestions here?

pritha-srivastava commented:
jenkins test make check arm64

pritha-srivastava commented:
jenkins test make check arm64

dang commented:
arm64 isn't required.  I think we can merge anyway?

mattbenjamin commented:
@dang @cbodley @alimaredia merging is trivially blocked because not approved by any reviewer?

ivancich commented on src/rgw/CMakeLists.txt:
```diff
@@ -116,6 +116,8 @@ set(librgw_common_srcs
   rgw_role.cc
   rgw_sal.cc
   rgw_sal_filter.cc
+  rgw_redis_driver.cc
+  rgw_ssd_driver.cc
```
Should `rgw_redis_driver.cc` be protectected with `if(WITH_RADOSGW_D4N)`? I just tried to build main and got:

/home/ivancich/ceph-work7/ceph/src/rgw/rgw_redis_driver.cc:2:10: fatal error: boost/redis/src.hpp: No such file or directory

ivancich commented on src/rgw/CMakeLists.txt:
```diff
@@ -148,6 +150,9 @@ set(librgw_common_srcs
   rgw_tracer.cc
   rgw_lua_background.cc
   rgw_data_access.cc
+  driver/d4n/d4n_directory.cc
+  driver/d4n/d4n_policy.cc
+  driver/d4n/rgw_sal_d4n.cc
```
Should `driver/d4n/*.cc` be protectected with `if(WITH_RADOSGW_D4N)`?
